pkg/customresourcestate/config.go

```diff
@@ -158,21 +166,46 @@ type ConfigDecoder interface {
 	Decode(v interface{}) (err error)
 }
 
-// FromConfig decodes a configuration source into a slice of customresource.RegistryFactory that are ready to use.
-func FromConfig(decoder ConfigDecoder) ([]customresource.RegistryFactory, error) {
-	var crconfig Metrics
+// FromConfig decodes a configuration source into a slice of `customresource.RegistryFactory` that are ready to use.
+func FromConfig(decoder ConfigDecoder, r *discovery.GVKMaps) ([]customresource.RegistryFactory, error) {
+	var customResourceConfig Metrics
 	var factories []customresource.RegistryFactory
 	factoriesIndex := map[string]bool{}
-	if err := decoder.Decode(&crconfig); err != nil {
+	if err := decoder.Decode(&customResourceConfig); err != nil {
 		return nil, fmt.Errorf("failed to parse Custom Resource State metrics: %w", err)
 	}
-	for _, resource := range crconfig.Spec.Resources {
+	// Cannot proceed ahead without an instantiated GVKMaps object (only for the
+	// first time, otherwise the cache will come up empty on the first run (in the
+	// <30s timespan).
+	for {
+		if !r.Heartbeat.IsZero() {
+			break
+		}
+	}
+	var resources []Resource
+	for _, resource := range customResourceConfig.Spec.Resources {
+		gotGVKs := r.ResolveGVK(discovery.GroupVersionKind(resource.GroupVersionKind))
+		if gotGVKs != nil {
+			var newResourceList []Resource
+			resource.wasResolved = true
+			for _, gvk := range gotGVKs {
+				newResource := resource
+				newResource.GroupVersionKind = GroupVersionKind(gvk)
+				newResourceList = append(newResourceList, newResource)
+			}
+			resources = append(resources, newResourceList...)
+		} else {
+			resources = append(resources, resource)
+		}
```

Comment:
Do we ever truncate GVKs? Like if a CRD is deleted from an apigroup, don't we need to clear the cache?

Reply:
The cache is refreshed (entirely) every 30s, so we should be fine there IIUC, since we do not have a dependency on CRDs now, but the `discovery` package itself that takes care of updating the cache for us.

Reply:
FromConfig is only called once or like whenever the config is updated, so the list of factories will not be up-to-date with the cache?

Reply:
For `t=0` to `t=30`, if, say, a `G**` config was provided, there's two cases. At the current time `t`, assume that the cache and factories are out-of-sync. 

NOTE: `Resource` below is a resource that satisfies the `G**` combination.

- Resource deletion: Default behaviour, as of now, is to not generate metrics for the resource that does not exist. So we'll basically keep serving metrics for the set of resolved GVKs sent back from the cache to KSM, but the ones that actually exist in the cluster.
- Resource addition: Metrics will be served after the cache refresh happens.

Reply:
Okay, so additionally caomparing the cache with the factories we have, and updating if need be, instead of waiting for the config-based reload to happen, and doing this check every 30s?

Reply:
If a CRD is deleted, the no metrics will be generated for it, but we'll still continue serving metrics all resources that match the `gvk` combination.

Reply:
I think we need to make sure there's a periodic sync between the cache and factories, instead of just the start, or when the config is edited.

Reply:
https://github.com/kubernetes/kube-state-metrics/commit/8320aeddb5ae3d4eb16d9d7d98b3a060846de1a1#diff-bf91d16f16c99a97ad09eb6587f27d8fc075a68825d396e035c8100f864a7582R177

Reply:
Cache is updated every [`discovery.Interval` seconds](https://github.com/kubernetes/kube-state-metrics/pull/1851/commits/cb95c2d24760336b7769a0309381ef1b6556ca72#diff-dc8fe36bd1edf1b4cd03bacbd94b02cd4226717fe7e1a6474c534f5b5db30227R162) now.